### PR TITLE
JAMES-2902 Sample integration test for SMTP HELO hook

### PR DESCRIPTION
**Pull request by chibenwa**
This intend to be used by contributors for contributors as a reference for implementing the other hooks integration tests. (HacktoberFest)

### CODE REVIEW

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.deletingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout`

test this please

**Comment by chibenwa** [comment]
```

[6baf0a71483a0aa73c1042ad678c1517a42303ad] [ERROR] Failures: 
[6baf0a71483a0aa73c1042ad678c1517a42303ad] [ERROR]   DistributedTaskManagerTest.listShouldBeEmptyWhenNoTasks 
[6baf0a71483a0aa73c1042ad678c1517a42303ad] Expecting empty but was:<[TaskExecutionDetails{taskId=TaskId{value=99571076-1fc9-4565-b392-36e53f8ff974}, type=TaskType{typeName=memory-reference-task}, additionalInformation=Optional.empty, status=IN_PROGRESS, submittedDate=2019-09-26T08:27:14.460Z[GMT], submittedNode=Hostname(foo), startedDate=Optional[2019-09-26T08:27:14.515Z[GMT]], ranNode=Optional[Hostname(foo)], completedDate=Optional.empty, canceledDate=Optional.empty, cancelRequestedNode=Optional.empty, failedDate=Optional[2019-09-26T08:27:14.515Z[GMT]]}]>
```

test this please

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/configuration/SmtpConfiguration.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
 import java.util.Optional;
 
 import org.apache.commons.io.IOUtils;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/configuration/SmtpConfiguration.java]: bad import

File: server/mailet/integration-testing/src/test/resources/smtpserver.xml [hunk 2]
```diff
@@ -44,6 +44,9 @@
         <handlerchain>
             <handler class="org.apache.james.smtpserver.fastfail.ValidRcptHandler"/>
             <handler class="org.apache.james.smtpserver.CoreCmdHandlerLoader"/>
+            {{#hooks}}
+            <handler class="{{hookFqcn}}"/>
+            {{/hooks}}
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/resources/smtpserver.xml]: you cannot generate a configuration object?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/extensions/SMTPHeloHooksTest.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        messageSender.connect(LOCALHOST_IP, jamesServer.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .sendMessage(FROM, TO);
+
+        awaitAtMostOneMinute.until(() -> resultChecker.getResults().size() > 0);
+        assertThat(resultChecker.getResults())
+            .containsExactly(Pair.of(RecordingHeloHook.class, DEFAULT_DOMAIN));
+    }
+
+    @Test
+    public void mailShouldBeWellDeliveredUponDeclinedHeloHook() throws Exception {
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/extensions/SMTPHeloHooksTest.java]: why? (maybe I don't understand what is a DeclinedHeloHook...)

File: server/mailet/integration-testing/src/test/resources/smtpserver.xml (hunk 2 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/resources/smtpserver.xml]: No, that's out of scope, and sadly not "easy" to do. (spoiler this means reworking concequently all Protocols (IMAP + SMTP + LMTP + FetchMail + ManageSieve).

Thank you inheritance.

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/extensions/SMTPHeloHooksTest.java (hunk 3 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/smtp/extensions/SMTPHeloHooksTest.java]: Decline => skipMe Hook

I know that is counter intuitive but you know, I cannot change everything at once on my spare time ;-)

**Comment by chibenwa** [comment]
```
[f709e2cdff766826f399e5b7536f4f95cb1e976c] [ERROR] error reading /root/.m2/repository/org/apache/commons/commons-lang3/3.9/commons-lang3-3.9.jar; zip file is empty
[f709e2cdff766826f399e5b7536f4f95cb1e976c] [ERROR] /james-project/server/container/lifecycle-api/src/main/java/org/apache/james/lifecycle/api/Startable.java:[19,1] cannot access org.apache.james.lifecycle.api
[f709e2cdff766826f399e5b7536f4f95cb1e976c] [ERROR]   zip END header not found
```

test this please

**Comment by chibenwa** [comment]
This work is merged

**Pull request closed without merging**